File: DataStructurePython/heap.py
class MaxHeap:
    def __init__(self):
        self.heap_array = []


    def insert(self, key):
        new_node = self.Node(key)
        self.heap_array.append(new_node)
        self.__percolate_up(len(self.heap_array) - 1)


    def __percolate_up(self, index: int):
        # save the bottom node
        bottom = self.heap_array[-1]
        # find the initial index value of parent
        parent = int((index - 1) / 2)
        # while parent's key is smaller than the new key
        while index > 0 and self.heap_array[parent].key < bottom.key:
            # parent node comes down
            self.heap_array[index] = self.heap_array[parent]
            index = parent # index moves up
            parent = int((parent - 1) / 2)
        # finally, insert newly added node into proper position
        self.heap_array[index] = bottom


    def remove_max(self) -> int:
        if not self.heap_array:
            return
        root = self.heap_array[0]
        self.heap_array[0] = self.heap_array[-1]
        del self.heap_array[-1]
        if self.heap_array:
            self.__perculate_down(0)
        return root.key


    def __perculate_down(self, index):
        top = self.heap_array[index]
        larger_child = -1 # larger child's index
        while index < int(len(self.heap_array) / 2):
            left_child = index * 2 + 1
            right_child = (index + 1) * 2
            # find which one is larger
            if right_child < len(self.heap_array) and self.heap_array[left_child].key < self.heap_array[right_child].key:
                larger_child = right_child
            else:
                larger_child = left_child
            # no need to go down any more
            if self.heap_array[larger_child].key <= top.key:
                break
            # move the nodes up
            self.heap_array[index] = self.heap_array[larger_child]
            # index goes down toward larger child
            index = larger_child
        # put top key into proper location to restore the heap
        self.heap_array[index] = top


    # Node
    class Node:
        def __init__(self, key):
            self.key = key

File: DataStructureSwift/Sources/DataStructureSwift/Heap.swift
//
//  Heap.swift
//  
//
//  Created by Yubo Qin on 11/30/22.
//

import Foundation

class MaxHeap<T: Comparable> {

    class Node<T> {
        let key: T
        init(key: T) {
            self.key = key
        }
    }

    private var heapArray: [Node<T>]

    init() {
        self.heapArray = []
    }

    func insert(_ key: T) {
        let node = Node(key: key)
        heapArray.append(node)
        percolateUp(index: heapArray.count - 1)
    }

    private func percolateUp(index: Int) {
        // save the bottom node
        guard let bottom = heapArray.last else {
            return
        }
        var index = index
        // find the initial index value of parent
        var parent = (index - 1) / 2
        // while parent's key is smaller than the new key
        while index > 0 && heapArray[parent].key < bottom.key {
            // parent node comes down
            heapArray[index] = heapArray[parent]
            // index moves up
            index = parent
            // continue with its parent
            parent = (parent - 1) / 2
        }
        // finally, insert newly added node into proper position
        heapArray[index] = bottom
    }

    func removeMax() -> Node<T>? {
        guard let root = heapArray.first else {
            return nil
        }
        if let last = heapArray.last {
            heapArray[0] = last
            heapArray.removeLast()
        }
        if !heapArray.isEmpty {
            perculateDown(index: 0)
        }
        return root
    }

    private func perculateDown(index: Int) {
        var index = index
        let top = heapArray[index]
        // index of larger child
        var largerChild = -1
        while index < heapArray.count / 2 {
            let leftChild = index * 2 + 1
            let rightChild = (index + 1) * 2
            // find which child is larger
            if rightChild < heapArray.count,
               heapArray[leftChild].key < heapArray[rightChild].key {
                largerChild = rightChild
            } else {
                largerChild = leftChild
            }
            // no need to go down any more
            if heapArray[largerChild].key < top.key {
                break
            }
            // move the nodes up
            heapArray[index] = heapArray[largerChild]
            // index goes down towards larger child
            index = largerChild
        }
        // put top key into proper location to restore the heap
        heapArray[index] = top
    }

}

func heapSort(data: [Int]) -> [Int] {
    let heap = MaxHeap<Int>()
    for i in data {
        heap.insert(i)
    }
    var ret: [Int] = []
    for _ in 0 ..< data.count {
        if let max = heap.removeMax() {
            ret.insert(max.key, at: 0)
        }
    }
    return ret
}

extension MaxHeap: CustomStringConvertible {
    var description: String {
        return "MaxHeap: \(heapArray)"
    }
}

extension MaxHeap.Node: CustomStringConvertible {
    var description: String {
        return "\(key)"
    }
}
